### ISSUE TITLE
[String Interpolation] Object key treating as html tag

### ISSUE THREAD

**Issue by indraraj26** [angular/angular#40902]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/core

### Is this a regression?

 Yes, the previous version in which this bug was not present was: No Idea

### Description

I am trying to show this object ` item = { "<Label>k__BackingField": "Aachi" };` on dom via string interpolation.
html
```
<!-- String interpolation is not working -->
<div>
	{{item["<Label>k__BackingField"]}}
</div>


<!-- this is also working -->
<div innerText="{{item['<Label>k__BackingField']}}">
</div>
<div [innerText]="item['<Label>k__BackingField']">
</div>

<!-- Property binding is working -->
<input type="text" [value]="item['<Label>k__BackingField']">
```
ts
```
 item = { "<Label>k__BackingField": "Testing" };
```

## 🔬 Minimal Reproduction

 
https://stackblitz.com/edit/angular-ivy-y5xtzu?file=src%2Fapp%2Fapp.component.html

## 🔥 Exception or Error
<pre><code>

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

11.0.8
</code></pre>

**Anything else relevant?**

**Comment by petebacondarwin** [comment]
I am confident that this is because our tokenizer does not distinguish between text and interpolations, so when it sees the `<Label>` in the text it treats it as the end of the text block and the beginning of a new HTML element.

**Comment by petebacondarwin** [comment]
The workaround is to split up the `<Label>` string in the interpolation: https://stackblitz.com/edit/angular-ivy-xemeuv?file=src%2Fapp%2Fapp.component.html:

```
{{item["<" + "Label>k__BackingField"]}}
```

**Comment by iRealNirmal** [comment]
@petebacondarwin it seems failing at lexer inside parser.

But you think we should modify rule  and allow to ignore html closing inside single quote of interpolation ?

**Comment by petebacondarwin** [comment]
The fix is not too easy because we would need the lexer to be able to understand more about text nodes than it currently does, yes?

**Comment by iRealNirmal** [comment]
Yes it's difficult and not only that also it's possible other devs might be using tag inside single quote. Consider even we do this change it would break things for them.

Number of devs using tag inside single quote will be quite higher compare to this scenario I believe we should not do this change.

**Comment by alxhub** [comment]
The rough plan for the compiler is to _eventually_ have a unified template and expression parser, that parses directly from string to an Angular AST and does not use a separate lexer/tokenizer. This is also hugely beneficial to the Language Service, as it enables much better error recovery when the template parsing operation isn't split across three separate parsers.

```
{{item["<Label>k__BackingField"]}}
```
should parse unambiguously, as `<` is not a special character within a quoted string in an expression context.

I don't _think_ this would be a breaking change. At least, I can't think of an example where a valid template today would result in a different parse against a unified "Angular Template Language".
